core/src/components/toggle/test/bottom-content/toggle.e2e.ts

```diff
@@ -0,0 +1,196 @@
+
+      expect(ariaDescribedBy).toBe(errorTextId);
+    });
+    test('input should have aria-invalid attribute when toggle is invalid', async ({ page }) => {
+      await page.setContent(
+        `<ion-toggle class="ion-invalid ion-touched" helper-text="Helper text" error-text="Error text">Label</ion-toggle>`,
+        config
+      );
+
+      const input = page.locator('ion-toggle input[type=checkbox]');
+
+      await expect(input).toHaveAttribute('aria-invalid');
+    });
+    test('input should not have aria-invalid attribute when toggle is valid', async ({ page }) => {
+      await page.setContent(`<ion-toggle helper-text="Helper text" error-text="Error text">Label</ion-toggle>`, config);
+
+      const input = page.locator('ion-toggle input[type=checkbox]');
+
+      await expect(input).not.toHaveAttribute('aria-invalid');
+    });
+    test('input should not have aria-describedby attribute when no hint or error text is present', async ({ page }) => {
+      await page.setContent(`<ion-toggle>Label</ion-toggle>`, config);
+
+      const input = page.locator('ion-toggle input[type=checkbox]');
+
+      await expect(input).not.toHaveAttribute('aria-describedby');
+    });
+  });
+});
+
+/**
+ * Rendering is different across modes
+ */
+configs({ modes: ['ios', 'md'], directions: ['ltr'] }).forEach(({ title, screenshot, config }) => {
+  test.describe(title('toggle: helper text rendering'), () => {
+    // Check the default label placement, end, and stacked
+    [undefined, 'end', 'stacked'].forEach((labelPlacement) => {
+      test(`${
+        labelPlacement ? `${labelPlacement} label - ` : ''
+      }should not have visual regressions when rendering helper text`, async ({ page }) => {
+        await page.setContent(
+          `<ion-toggle ${
+            labelPlacement ? `label-placement="${labelPlacement}"` : ''
+          } helper-text="Helper text">Label</ion-toggle>`,
+          config
+        );
+
+        const bottomEl = page.locator('ion-toggle');
+        await expect(bottomEl).toHaveScreenshot(
+          screenshot(`toggle-helper-text${labelPlacement ? `-${labelPlacement}` : ''}`)
+        );
+      });
+
+      test(`${
+        labelPlacement ? `${labelPlacement} label - ` : ''
+      }should not have visual regressions when rendering helper text with wrapping text`, async ({ page }) => {
+        await page.setContent(
+          `<ion-toggle ${
+            labelPlacement ? `label-placement="${labelPlacement}"` : ''
+          } helper-text="Helper text helper text helper text helper text helper text helper text helper text helper text helper text">Label</ion-toggle>`,
+          config
+        );
+
+        const bottomEl = page.locator('ion-toggle');
+        await expect(bottomEl).toHaveScreenshot(
+          screenshot(`toggle-helper-text${labelPlacement ? `-${labelPlacement}` : ''}-wrapping`)
+        );
+      });
+    });
+  });
+
+  test.describe(title('toggle: error text rendering'), () => {
```

Comment:
Why does error text not get a check for the end or fixed placement?

Reply:
[Answered](https://github.com/ionic-team/ionic-framework/pull/30140#discussion_r1962447834)